### PR TITLE
New Feature - Nullish Coalescing (?? operator)

### PR DESCRIPTION
**Pull request by kisbg**
Signed-off-by: bence gabor kis <kisbg@inf.u-szeged.hu>

### CODE REVIEW

File: src/parser/ast/AssignmentExpressionNullishCoalescingNode.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+#ifndef AssignmentExpressionNullishCoalescingNode_h
+#define AssignmentExpressionNullishCoalescingNode_h
+
+#include "ExpressionNode.h"
+#include "AssignmentExpressionSimpleNode.h"
+
+namespace Escargot {
+
+class AssignmentExpressionNullishCoalescingNode : public ExpressionNode {
```
Comment (clover2123) [inline src/parser/ast/AssignmentExpressionNullishCoalescingNode.h:28]: What is this `AssignmentExpressionNullishCoalescingNode` for?
It seems that nullish coalescing (?? operator) is only applicable for binary expression, not for assignment case. (`??=` operatoration).
Correct me if I'm wrong.

File: src/parser/esprima_cpp/esprima.cpp [hunk 2]
```diff
@@ -2521,6 +2521,14 @@ class Parser {
             ASTNode right = this->isolateCoverGrammar(builder, &Parser::parseExponentiationExpression<ASTBuilder>);
             expr = this->finalize(this->startNode(startToken), builder.createBinaryExpressionExponentiationNode(left, right));
         }
+        if (this->match(NullishCoalescing)) {
+            this->nextToken();
+            this->context->isAssignmentTarget = false;
```
Comment (clover2123) [inline src/parser/esprima_cpp/esprima.cpp:2526]: Should nullish operation be handled in `parseExponentiationExpression`?
Then the name of the function should also be fixed to represent its role.

**Comment by clover2123** [comment]
Our test262 is newly updated (#688).
So please rebase and update related test262 TCs according to this. Thanks.
NOTE) You can easily make exclude list by running the following script
```sh
tools/test/test262/make_excludelist.py --arch={x86|x86_64} --engine={ESCARGOT_PATH}
```

File: src/interpreter/ByteCodeInterpreter.cpp [hunk 3]
```diff
@@ -1559,6 +1572,7 @@ NEVER_INLINE Value ByteCodeInterpreter::exponentialOperation(ExecutionState& sta
     return Value(pow(base, exp));
 }
 
+
```
Comment (clover2123) [inline src/interpreter/ByteCodeInterpreter.cpp:1575]: Please remove this added line.

File: src/interpreter/ByteCodeGenerator.cpp [hunk 4]
```diff
@@ -611,7 +617,8 @@ ByteCodeBlock* ByteCodeGenerator::generateByteCode(Context* c, InterpretedCodeBl
             case BinarySignedRightShiftOpcode:
             case BinaryUnsignedRightShiftOpcode:
             case BinaryInOperationOpcode:
-            case BinaryInstanceOfOperationOpcode: {
+            case BinaryInstanceOfOperationOpcode:
+            case BinaryExponentiationOpcode: {
```
Comment (clover2123) [inline src/interpreter/ByteCodeGenerator.cpp:621]: `BinaryExpressionNullishCoalescing` bytecode does not need stack index alignment?

File: src/interpreter/ByteCodeInterpreter.cpp [hunk 5]
```diff
@@ -602,6 +602,19 @@ Value ByteCodeInterpreter::interpret(ExecutionState* state, ByteCodeBlock* byteC
             NEXT_INSTRUCTION();
         }
 
+        DEFINE_OPCODE(JumpIfUndefinedorNull)
+            :
+        {
+            JumpIfUndefinedorNull* code = (JumpIfUndefinedorNull*)programCounter;
+            ASSERT(code->m_jumpPosition != SIZE_MAX);
+            if (!registerFile[code->m_registerIndex].isUndefinedOrNull()) {
+                programCounter = code->m_jumpPosition;
```
Comment (clover2123) [inline src/interpreter/ByteCodeInterpreter.cpp:611]: BTW if this bytecode jumps only when the register value is not undefined nor null, the name of this bytecode should be changed according to its operation such as `JumpIfNotUndefinedNorNull`.

Comment (clover2123) [reply src/interpreter/ByteCodeInterpreter.cpp:611]: One more thing.
It seems that this new bytecode needs jump position alignment like other Jump* bytecodes cases bekiw.
https://github.com/Samsung/escargot/blob/1ab89947116a1e2f53011e828bbcef8cc038c875/src/interpreter/ByteCodeGenerator.cpp#L513-L518

**Comment by clover2123** [comment]
@kisbg 
I kindly ask you to communicate more actively with reviewers.
We think that communication & discussion is really important thing in development process,
but you didn't answer any comment of reviewers in this pr and it makes review very difficult.
(We don't know what you do think about each review comment or what your intention is)
Please consider it next.

**Comment by kisbg** [comment]
@clover2123  I apologize for the problem is caused. I promised in the future i going to be more active in the review process.

File: src/interpreter/ByteCodeInterpreter.cpp (hunk 5 above)
Comment (kisbg) [reply src/interpreter/ByteCodeInterpreter.cpp:611]: Can you elaborate for me? What you mean by needs 'jump position alignment' ?  I used the other bytecode jumps as a template. And im not familiar this part of the code.

Comment (clover2123) [reply src/interpreter/ByteCodeInterpreter.cpp:611]: `m_jumpPosition` should be fixed to indicate the real address of jump target.
You can find that `cd->m_jumpPosition = cd->m_jumpPosition + codeBase;` statement does this job.
When the bytecode is generated for the first time, it only calculates the offset of jump target. 
So we fix it later as the above code part.
Also register index is aligned correctly if it is necessary too.

File: src/interpreter/ByteCodeGenerator.cpp [hunk 6]
```diff
@@ -516,6 +516,12 @@ ByteCodeBlock* ByteCodeGenerator::generateByteCode(Context* c, InterpretedCodeBl
                 ASSIGN_STACKINDEX_IF_NEEDED(cd->m_registerIndex, stackBase, stackBaseWillBe, stackVariableSize);
                 break;
             }
+            case JumpIfNotUndefinedNorNullOpcode: {
```
Comment (kisbg) [inline src/interpreter/ByteCodeGenerator.cpp:519]: @clover2123  Sorry for the inconvenience, but at the moment the code looks just like all the others, so im not sure what you want to fix here. Can you write the correct solution?

Comment (clover2123) [reply src/interpreter/ByteCodeGenerator.cpp:519]: Oh,, It was my mistake. Maybe I mistakenly omitted some part of your update.
Sorry for the confusion.

**Pull request merged**